src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.cs

```diff
@@ -501,9 +515,39 @@ private bool IsFirstTokenOnLine(TokenData tokenData1, TokenData tokenData2)
             return this.GetTriviaData(tokenData1, tokenData2).SecondTokenIsFirstTokenOnLine;
         }
 
+        /// <summary>
+        /// Gets approximated index bounds for a token with specified position, which will definitely contain
+        /// the token, if it is present in the stream.
+        /// </summary>
+        private void GetTokenIndexBounds(int position, out int lowerBound, out int upperBound)
+        {
+            float avgLength = this.AverageTokenLength;
+            Debug.Assert(avgLength > 0);
+            // make an initial guess about the token position
+            lowerBound = upperBound = Bounded((int)((position - _tokens[0].FullSpan.Start) / avgLength), min: 0, max: _tokens.Length - 1);
+
+            // decrease lower bound until the desired position is strictly after it
+            while (_tokens[lowerBound].FullSpan.Start >= position && lowerBound > 0)
+            {
+                int estimatedError = (int)((_tokens[lowerBound].FullSpan.Start - position) / avgLength);
+                lowerBound = Math.Max(0, lowerBound - Math.Max(1, estimatedError));
```

Comment:
i think this line is totally reasonable, but i also think it warrants a comment to help people along.  i.e. "we want to make sure we at least reduce the lower bound by 1, but we can't even go past the start of the sequence of tokens".  Up to you.

Reply:
I think this would be that kind of comment, that would not be kept in sync with code, if code changes later on.